src/middlewared/middlewared/plugins/acme_protocol.py

```diff
@@ -0,0 +1,380 @@
+            {
+                'uri': register.uri,
+                'tos': register.terms_of_service,
+                'new_account_uri': directory.newAccount,
+                'new_nonce_uri': directory.newNonce,
+                'new_order_uri': directory.newOrder,
+                'revoke_cert_uri': directory.revokeCert,
+                'directory': data['acme_directory_uri']
+            }
+        )
+
+        # Save registration body
+        self.middleware.call_sync(
+            'datastore.insert',
+            'system.acmeregistrationbody',
+            {
+                'contact': register.body.contact[0],
+                'status': register.body.status,
+                'key': key.json_dumps(),
+                'acme': registration_id
+            }
+        )
+
+        return self.middleware.call_sync(f'{self._config.namespace}._get_instance', registration_id)
+
+
+class DNSAuthenticatorService(CRUDService):
+
+    class Config:
+        namespace = 'dns.authenticator'
+        datastore = 'system.dnsauthenticator'
+
+    def __init__(self, *args, **kwargs):
+        super(DNSAuthenticatorService, self).__init__(*args, **kwargs)
+        self.schemas = DNSAuthenticatorService.get_authenticator_schemas()
+
+    @staticmethod
+    def get_authenticator_schemas():
+
+        return {
+            f_n[len('update_txt_record_'):]: [
+                Str(arg, required=True)
+                for arg in list(getattr(DNSAuthenticatorService, f_n).__code__.co_varnames)
+                [4: getattr(DNSAuthenticatorService, f_n).__code__.co_argcount]
+            ]
+            for f_n in [
+                func for func in dir(DNSAuthenticatorService)
+                if callable(getattr(DNSAuthenticatorService, func))
+                and func.startswith('update_txt_record_')
+            ]
+        }
+
+    async def common_validation(self, data, schema_name):
+        verrors = ValidationErrors()
+        if data['authenticator'] not in self.schemas:
+            verrors.add(
+                f'{schema_name}.authenticator',
+                f'System does not support {data["authenticator"]} as an Authenticator'
+            )
+        else:
+            verrors = validate_attributes(self.schemas[data['authenticator']], data)
+
+        if verrors:
+            raise verrors
+
+    async def schema_choices(self):
+        return [
+            {
+                'name': key,
+                'title': key.replace('_', ' ').capitalize(),
+                'credentials_schema': [
+                    {
+                        'property': v.name,
+                        'schema': {
+                            '_required_': False,
+                            'title': v.name.replace('_', ' ').capitalize(),
+                            'type': 'string'
+                        }
+                    }
+                    for v in value
+                ]
+            }
+            for key, value in self.schemas.items()
+        ]
```

Comment:
That does not seems like a valid JSON schema, why is it used for exactly and why we can just to .to_json_schema for each?

Also is it a public method? It neeeds @accepts() if so.